text/000-buildpacks-contribute-default-process-type.md

```diff
@@ -0,0 +1,133 @@
+
+In line with the existing philosophy, this RFC proposes that the `default-process` type from later buildpacks should override any `default-process` types from earlier buildpacks UNLESS `override = false`. The user-provided `-process-type` flag would override any default process type contributed by buildpacks.
+
+Buildpacks should not be able to declare a process type as the default if it is not a process type that they define (see [edge cases](#edge-cases)). Additionally, if a buildpack attempts to redefine a process type that is declared as the default by an earlier buildpack, the default designation should be cleared unless explicitly set by the later buildpack.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+
+- Buildpacks could stomp on each other, each trying to declare different default process types. This might be confusing.
+- Changing the way things work now could surprise some people.
+- One more thing for buildpack authors to know and care about.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- What other designs have been considered? Keep things as they are.
+- Why is this proposal the best? Easier for end users! It simplifies the build process (need to provide one fewer flag, don't need to rebuild if process types are not known ahead of time).
+- What is the impact of not doing this? A moderate amount of frustration.
+
+# Prior Art
+[prior-art]: #prior-art
+
+Discuss prior art, both the good and bad.
+
+`pack build` has a `--default-process` flag. If this flag is not provided by the user, `pack` will provide `-process-type web` to the lifecycle. If `web` is not a valid process type, the lifecycle will warn and set the image entrypoint to `/cnb/lifecycle/launcher`, which will require the user to specify the desired process in the command when running the app image. This has the advantage of "just working" for web apps. However for non-web apps, the user must know what process types exist and pass the desired default explicitly when building the app, otherwise they will get an image that must be configured at runtime.
+
+The lifecycle if not provided any `-process-type` will set the default process type to the only process type (if there is only one), otherwise it will warn and set the image entrypoint to `/cnb/lifecycle/launcher` as described above. This has the advantage of "just working" for single-process apps. However this requires the lifecycle to be "opinionated" and may be surprising for some users. See [spec PR](https://github.com/buildpacks/spec/pull/137) for the proposed removal of this behavior which should go hand-in-hand with this RFC.
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- Should `pack` continue to pass `-process-type web`? It could inspect the buildpack API and only pass the flag if the API version is below 0.X. 
+- What about buildpacks implementing different buildpack APIs that need to work together?
```

Comment:
Thanks for marking this question. I'm interested in what people think should happen here. Since this functionality doesn't exist in previous APIs I believe there are only two options:
* we assume buildpacks in a previous API provide no default process types and this RFC would function as stated. This is clean but may come as a surprise
* we assume buildpacks that define a `web` process type intend for it to be default.

Reply:
I believe this is covered [here](https://github.com/buildpacks/rfcs/pull/110/files#diff-00a2fb620563b974f876628cc19e746a4a96606f9dea0fd58dc1bb13f010b4abR166) and [here](https://github.com/buildpacks/rfcs/pull/110/files#diff-00a2fb620563b974f876628cc19e746a4a96606f9dea0fd58dc1bb13f010b4abR105).